### PR TITLE
bpo-40550: Fix time-of-check/time-of-action issue in subprocess.Popen.send_signal.

### PR DESCRIPTION
**Pull request by FFY00**
All the implementation details have been explained the bug, please check it out.

https://bugs.python.org/issue40550

### CODE REVIEW

File: Lib/subprocess.py [hunk 1]
```diff
@@ -2061,7 +2061,11 @@ def send_signal(self, sig):
             # The race condition can still happen if the race condition
             # described above happens between the returncode test
             # and the kill() call.
-            os.kill(self.pid, sig)
+            try:
+                os.kill(self.pid, sig)
+            except ProcessLookupError as e:
+                if e.errno != 3:
```
Comment (ZackerySpytz) [inline Lib/subprocess.py]: Please do not use a hardcoded 3 here. Use the symbolic name.

File: Misc/NEWS.d/next/Library/2020-05-08-21-30-54.bpo-40550.i7GWkb.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix time-of-check/time-of-action issue in multiprocessing.Popen.send_signal.
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2020-05-08-21-30-54.bpo-40550.i7GWkb.rst]: "multiprocessing.Popen.send_signal" is not valid.

**Review by ZackerySpytz — COMMENTED** [review]
I'm not sure if this PR will be accepted, but I left some comments.

Also, this PR should probably have a unit test.

File: Misc/NEWS.d/next/Library/2020-05-08-21-30-54.bpo-40550.i7GWkb.rst (hunk 2 above)
Comment (FFY00) [reply Misc/NEWS.d/next/Library/2020-05-08-21-30-54.bpo-40550.i7GWkb.rst]: Ups, sorry.

**Comment by FFY00** [comment]
This is a race condition, there's no way to reliably have a unit test.

**Comment by FFY00** [comment]
This fixes a race condition, there's no way to reliably have a unit test.

The process needs to exit between the return code check and the action itself.

**Comment by remilapeyre** [comment]
I think you could write a unit test by synchronising the threads so the race condition occurs.

You can mock `os.kill()` so that it waits for a `threading.Event` before continuing, then wait in another thread for the process to die. Once it's dead, you can release the first thread by setting the event and the `kill()` should raise a `ProcessLookupError`.

File: Lib/subprocess.py [hunk 3]
```diff
@@ -2061,7 +2061,13 @@ def send_signal(self, sig):
             # The race condition can still happen if the race condition
             # described above happens between the returncode test
             # and the kill() call.
-            os.kill(self.pid, sig)
+            try:
+                os.kill(self.pid, sig)
+            except ProcessLookupError as e:
+                # supress the process not found error in case the race
+                # condition happens
```
Comment (remilapeyre) [inline Lib/subprocess.py]:
```suggestion
                # condition happens, see bpo-40550
```

**Review by remilapeyre — APPROVED** [review]
Hi @FFY00, I checked the test and it simulates the race condition successfully 👍 

Try not to force-push during reviews thought as it makes it harder to see what has been updated since the last time one viewed that changes.

**Comment by FFY00** [comment]
You can click on `force-pushed` (in ` @FFY00 FFY00 force-pushed the FFY00:bpo-40550 branch from 649b00f to 49e7004 5 hours ago `), it should land you in this link: https://github.com/python/cpython/compare/649b00f3308818d88e237083600c1955162fdadd..49e7004bc963966bfb2f264d6022ef4477f21bcc. You can see the changes perfectly. It shows you the changes just as same as if I were to push a different commit, but without dirtying up the history. This might not be a problem in simple PRs like this but it makes a difference when you have more than 1 commit. And makes a huge difference when you are changing a lot of code, since you won't be able to look at one clean commit, you will need to look at the original commit and the fix. For these reasons, I believe force pushing to be good practice.

**Comment by FFY00** [comment]
Sorry, after writing all that I messed up the git history :confused:

You can still view the correct diff here: https://github.com/python/cpython/compare/49e7004bc963966bfb2f264d6022ef4477f21bcc..3d7dfbc1c19280dd19693d8d8e0574590c0c1bee

**Comment by FFY00** [comment]
@gpshead do you have a minute to review this?

File: Lib/subprocess.py [hunk 4]
```diff
@@ -2061,7 +2061,13 @@ def send_signal(self, sig):
             # The race condition can still happen if the race condition
             # described above happens between the returncode test
             # and the kill() call.
-            os.kill(self.pid, sig)
+            try:
+                os.kill(self.pid, sig)
+            except ProcessLookupError as e:
+                # supress the process not found error in case the race
+                # condition happens, see bpo-40550
+                if e.errno != errno.ESRCH:
```
Comment (gpshead) [inline Lib/subprocess.py]: This is harmless but redundant.  ProcessLookupError only occurs when `e.errno == ESRCH`.

https://github.com/python/cpython/blob/v3.10.0a2/Objects/exceptions.c#L2653

(i'll submit an edit turning this into a mere `pass`)

**Review by gpshead — APPROVED** [review]

File: Lib/subprocess.py (hunk 4 above)
Comment (FFY00) [reply Lib/subprocess.py]: Ah, gotcha!

**Pull request merged**